tracing/tracing_test.go

```diff
@@ -112,11 +109,83 @@ func TestOverrideTraceResourceNameExperimental(t *testing.T) {
 
 	span, spanCtx := tracer.StartSpanFromContext(ctx, "test", tracer.ResourceName("UnitTest"))
 	defer span.Finish()
-	os.Setenv(internal.ExperimentalTracingEnabled, "true")
-	defer func() {
-		os.Setenv(internal.ExperimentalTracingEnabled, "")
-	}()
+	t.Setenv(internal.ExperimentalTracingEnabled, "true")
 	err = OverrideTraceResourceName(spanCtx, newRes)
 
 	assert.Nil(t, err)
 }
+
+func TestStartChildSpan(t *testing.T) {
+	// Start Datadog tracer, so that we don't create NoopSpans.
+	_ = mocktracer.Start()
+
+	type args struct {
+		spanInCtx    bool
+		experimental bool
+	}
+	tests := []struct {
+		name string
+		args args
+	}{
+		{
+			name: "legacy without span",
+			args: args{
+				spanInCtx:    false,
+				experimental: false,
+			},
+		},
+		{
+			name: "legacy with span",
+			args: args{
+				spanInCtx:    true,
+				experimental: false,
+			},
+		},
+		{
+			name: "experimental without span",
+			args: args{
+				spanInCtx:    false,
+				experimental: true,
+			},
+		},
+		{
+			name: "experimental with span",
+			args: args{
+				spanInCtx:    true,
+				experimental: true,
+			},
+		},
+	}
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Setenv(internal.ExperimentalTracingEnabled, fmt.Sprintf("%v", tt.args.experimental))
+
+			ctx := context.Background()
+			if tt.args.spanInCtx {
+				if tt.args.experimental {
+					span, spanCtx := tracer.StartSpanFromContext(ctx, "test", tracer.ResourceName("UnitTest"))
+					defer span.Finish()
+					ctx = spanCtx
+				} else {
+					span, spanCtx := tracer.StartSpanFromContext(ctx, "test", tracer.ResourceName("UnitTest"))
+					defer span.Finish()
+					extCtx := internal.ExtendedContextWithMetadata(spanCtx, internal.TraceContextKey{}, TraceDetails{DatadogSpan: span})
```

Comment:
It's great to see the capability for nesting now available.
However, I have a couple of suggestions:

1. **Flexibility for Span Types**: There might be cases where a sequence of spans, rather than nesting, is required, as depicted in the screenshot. Tje library must allow the consumer to determine the type of span needed, reinforcing that the choice of span structure isn't dictated by the library but by the use case.
2. **Code Duplication**: To avoid redundancy, we should consider abstracting the following repetitive code block into a function - as was added generic. This approach would prevent each project from having to implement its own versions to handle this common pattern:
```go
span, spanCtx := tracer.Start(ctx, "spanName")
// do work
span.Finish() // or defer span.Finish()
```

By wrapping this into a function, we ensure consistency and simplicity across projects.

Additionally, as Soumesh pointed out, it would be beneficial to ensure that spans are closed with any relevant error information, providing better error context and tracing.

Reply:
Resolving the conversation due to inactivity.